### PR TITLE
[AAP-77766] Publish open source Jewel container image to GHCR

### PR DESCRIPTION
**Pull request by Dostonbek1**
## Description

- **What is being changed?** Adds a GitHub Actions workflow that builds and publishes a Jewel container image to GHCR (`ghcr.io/ansible/jewel`) on every push to `devel`. Also makes the platform UI inclusion optional in the Dockerfile via a build arg, and documents image usage in the README.
- **Why is this change needed?** Jewel currently has no published builds, making it difficult for open source contributors to use the project without building from scratch. AWX and EDA already publish upstream images — Jewel should follow suit.
- **How does this change address the issue?** Introduces a CI workflow modeled after EDA Server and AWX patterns, using GHCR with the built-in `GITHUB_TOKEN` (no custom secrets needed).

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update
- [x] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

### Prerequisites
- Push access to a fork of `ansible/jewel`

### Steps to Test
1. Merge to `devel` — the workflow runs automatically on push
2. Check the Actions tab for the "Build and Publish Image" workflow
3. Verify the image is available at `ghcr.io/ansible/jewel:latest`
4. Pull and run: `docker pull ghcr.io/ansible/jewel:latest`

### Expected Results
- Workflow completes successfully and pushes multi-arch (amd64/arm64) image
- Image is tagged with `latest`, `sha-<short>`, and `sha-<long>`
- Image does NOT include the platform UI by default
- Building with `--build-arg INCLUDE_UI=true` includes the UI

### Design Decisions
- **GHCR over Quay.io** — aligns with AWX's approach and signals upstream/devel. No custom robot account secrets needed.
- **`latest` tag** — rolling tag updated on every push to `devel`, plus immutable SHA-based tags for pinning.
- **`INCLUDE_UI=false` by default** — per acceptance criteria, the UI reference is removed by default with a build arg to opt in.
- **Fork guard** — `if: github.repository == 'ansible/jewel'` prevents the workflow from running on forks.

### Required Actions
- [ ] Requires coordination with other teams
  - Community team should be informed once the image is live
  - GHCR package visibility may need to be set to "public" by an org admin after the first publish

---
**Note:** This PR was developed with assistance from Claude AI assistant.

## Summary by CodeRabbit

- **New Features**
  - Added an automated “Build and Publish Image” workflow on the `devel` branch.
  - Publishes multi-architecture container images (AMD64 and ARM64) to GHCR with `latest` and commit-based tags (`sha-<short>` / `sha-<full>`).
  - Introduced image build targets to optionally include the platform UI (default builds use a placeholder).

- **Documentation**
  - Added a “Container Image” section with pull instructions, available tag formats, and local build guidance (including optional UI inclusion).

### CODE REVIEW

File: .github/workflows/build-image.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ansible/jewel
+
+jobs:
+  build-and-push-image:
+    if: github.repository == 'ansible/jewel'
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
```
Comment (jay-steurer) [inline .github/workflows/build-image.yml:22]: **Critical — Supply chain risk: Actions pinned by mutable version tags**

All actions use mutable tags (`@v4`, `@v3`, etc.). If an upstream action repo is compromised, arbitrary code can be injected into this pipeline (ref: `tj-actions/changed-files` incident, March 2025).

Pin every action to its full commit SHA with a version comment:
```yaml
- uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
- uses: docker/login-action@74a5d142397b4f367a81961aca4a556e4ec26871 # v3.4.0
- uses: docker/setup-qemu-action@29109295f81e9208d7d86ff1c6c12d2833863392 # v3.6.0
- uses: docker/setup-buildx-action@b5ca514318bd6ebac0fb2aedd5d36ec1b5c232a2 # v3.10.0
- uses: docker/metadata-action@902fa8ec7d6ecbf8d84d538b9b233a880e428804 # v5.7.0
- uses: docker/build-push-action@263435318d21b8e681c14492fe198571cfb76f00 # v6.18.0
```
(Verify exact SHAs against current releases.)

File: .github/workflows/build-image.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          tags: |
+            type=raw,value=latest
+            type=sha
+            type=sha,format=long
```
Comment (jay-steurer) [inline .github/workflows/build-image.yml:53]: **Important — `INCLUDE_UI=true` path is never tested in CI**

The workflow hardcodes `INCLUDE_UI=false` with no override mechanism. The `INCLUDE_UI=true` path is documented in the README but never exercised in CI — if the conditional copy logic breaks, no CI run will catch it.

Consider adding a `workflow_dispatch` trigger with an input parameter, or a second job that builds (no push) with `INCLUDE_UI=true` to validate both paths.

File: .github/workflows/build-image.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v6
```
Comment (jay-steurer) [inline .github/workflows/build-image.yml]: **Suggestion — No smoke test before pushing to `latest`**

If a broken commit lands on `devel`, the `latest` tag gets overwritten with a broken image. There is no validation between build and push.

Consider a build-then-test-then-push pattern: build with `push: false` + `load: true`, run a quick smoke test (`docker run --rm jewel:test python -c "import aap_gateway_api"`), then rebuild multi-arch with `push: true`.

File: README.md [hunk 4]
```diff
@@ -17,6 +17,29 @@ Jewel provides the source code for the Gateway that connects Ansible services an
 > The project name, Jewel, refers to a precious stone known as the Eye of the Sea from the same novel that originated the word Ansible, Ursula K. Le Guin's *Rocannon's World*.
 > Ansible communicates across distance; Jewel is what everything converges on.
 
+## Container Image
+
+A pre-built container image is published to the GitHub Container Registry on every push to the `devel` branch. The published image does **not** include the platform UI.
+
```
Comment (jay-steurer) [inline README.md:23]: **Important — README does not mention the published image excludes the platform UI**

A developer who pulls this image has no indication it ships without the UI. They will get a 404 in the browser with no explanation.

Suggested addition after the pull command:
> **Note:** The published image does not include the platform UI. To build an image with the UI included, see the local build instructions below.

File: tools/docker/Dockerfile [hunk 5]
```diff
@@ -91,11 +91,13 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
 ENV PATH=/opt/aap_gateway/venv/bin:$PATH
 ENV PYTHONPATH=/opt/aap_gateway/src
```
Comment (jay-steurer) [inline tools/docker/Dockerfile:93]: **Suggestion — Add a comment explaining the conditional UI inclusion**

The old comment explaining why this block is near the end of the Dockerfile was removed. The new code is more complex and the staging approach is non-obvious. Suggested:
```dockerfile
# Platform UI inclusion is optional. CI images (GHCR) are built without it.
# Because COPY --from cannot be conditional, we stage files in /tmp and
# selectively copy them. Keep this near the end — the upstream image changes frequently.
```

File: tools/docker/Dockerfile [hunk 6]
```diff
@@ -91,11 +91,13 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
...
 
-# Pull in the platform-ui, this can change frequently so it should be one of the last things we try and do
-COPY --from=quay.io/ansible/platform-ui:latest /usr/share/nginx/html /opt/aap_gateway/platform_ui
+RUN mkdir -p /opt/aap_gateway/platform_ui
```
Comment (jay-steurer) [inline tools/docker/Dockerfile]: **Critical — Empty `platform_ui` directory causes silent 404 for all browser requests**

When `INCLUDE_UI=false`, this directory exists but is empty. The nginx config (`tools/configs/nginx-gateway.conf:48-62`) serves from this path with `try_files $uri /index.html =404`. Since `index.html` doesn't exist, every browser request returns a raw 404 with no explanation.

Users pulling the GHCR image and opening it in a browser will think the container is broken.

Fix: add a placeholder `index.html` in the `else` branch that says the image is API-only and how to rebuild with `--build-arg INCLUDE_UI=true`.

File: tools/docker/Dockerfile [hunk 7]
```diff
@@ -91,11 +91,13 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
...
 
 USER 1000
```
Comment (jay-steurer) [inline tools/docker/Dockerfile:104]: **Important — Platform-UI image always pulled even when `INCLUDE_UI=false`**

This `COPY --from` is unconditional — Docker pulls the entire `quay.io/ansible/platform-ui:latest` image on every build regardless of `INCLUDE_UI`. This means:
- Builds with `INCLUDE_UI=false` fail if quay.io is unreachable (defeating the purpose of making UI optional)
- The files leak into a recoverable image layer even after the `rm -rf`
- Unnecessary network traffic and build time

Since the workflow uses Buildx (BuildKit), consider `RUN --mount=from=` to avoid a persistent layer, or restructure as a multi-stage build with a conditional target.

Also: `platform-ui:latest` is a mutable tag — consider pinning to a digest for reproducible builds.

File: tools/docker/Dockerfile [hunk 8]
```diff
@@ -91,11 +91,13 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
...
 EXPOSE 8000
```
Comment (jay-steurer) [inline tools/docker/Dockerfile:105]: **Important — `INCLUDE_UI` build arg has no validation; typos are silently ignored**

The check `[ "${INCLUDE_UI}" = "true" ]` is an exact string match. Values like `True`, `TRUE`, `yes`, `1`, or a typo like `ture` all silently evaluate to false — the UI is excluded with no warning.

Add validation:
```dockerfile
RUN if [ "${INCLUDE_UI}" != "true" ] && [ "${INCLUDE_UI}" != "false" ]; then \
      echo "ERROR: INCLUDE_UI must be 'true' or 'false', got '${INCLUDE_UI}'" && exit 1; \
    fi
```

**Review by jay-steurer — COMMENTED** [review]
## PR Review Summary

**Critical (2)** · **Important (3)** · **Suggestions (4)**

See inline comments for details. The main blockers before merge are:
1. Empty `platform_ui` directory silently serves 404 to all browser requests
2. GitHub Actions pinned by mutable version tags (supply chain risk)

**Strengths:** Fork guard, scoped permissions, multi-platform build, accurate README tag table, clear workflow step names.

File: tools/docker/Dockerfile [hunk 9]
```diff
@@ -91,11 +91,13 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
...
```
Comment (jay-steurer) [inline tools/docker/Dockerfile:106]: **Suggestion — `cp -r *` silently skips dotfiles (behavioral regression)**

The shell glob `*` does not match files beginning with `.` (e.g., `.well-known`, `.htaccess`). The original `COPY --from` copied everything including hidden files.

Fix: use `cp -a` with trailing dot syntax:
```dockerfile
cp -a /tmp/platform_ui/. /opt/aap_gateway/platform_ui/
```

File: .github/workflows/build-image.yml (hunk 1 above)
Comment (Dostonbek1) [reply .github/workflows/build-image.yml:22]: Fixed — all actions are now pinned to full commit SHAs with version comments (e.g. `actions/checkout@11d5960a...  # v4.4.0`).

File: .github/workflows/build-image.yml (hunk 2 above)
Comment (Dostonbek1) [reply .github/workflows/build-image.yml:53]: This was refactored — the `INCLUDE_UI` build arg no longer exists. The Dockerfile now uses named build targets: `jewel` (no UI, what CI publishes) and `jewel-ui` (extends `jewel` with a single `COPY --from`). The `jewel-ui` target is just one COPY instruction, so there's not much that can break independently.

File: .github/workflows/build-image.yml (hunk 3 above)
Comment (Dostonbek1) [reply .github/workflows/build-image.yml]: Good idea — deferring for a follow-up. The main complication is that Buildx multi-arch builds can't `load` into local Docker, so we'd need to build twice (once single-arch for testing, once multi-arch for push). Worth adding once we have more experience with the pipeline.

File: README.md (hunk 4 above)
Comment (Dostonbek1) [reply README.md:23]: This was already addressed in the previous push — the README now says: "The published image does **not** include the platform UI."

File: tools/docker/Dockerfile (hunk 5 above)
Comment (Dostonbek1) [reply tools/docker/Dockerfile:93]: Fixed — added a comment explaining the `jewel` / `jewel-ui` build target pattern and why UI inclusion is split into a separate stage.

File: tools/docker/Dockerfile (hunk 6 above)
Comment (Dostonbek1) [reply tools/docker/Dockerfile]: Great catch — fixed. The `jewel` target now writes a placeholder `index.html` that tells users this is an API-only image and how to rebuild with `--target jewel-ui` to include the UI.

File: tools/docker/Dockerfile (hunk 7 above)
Comment (Dostonbek1) [reply tools/docker/Dockerfile:104]: Already resolved — the Dockerfile was refactored to use named build targets instead of the `INCLUDE_UI` arg. When building with `--target jewel`, Docker never encounters the `COPY --from=quay.io/ansible/platform-ui` instruction (it's in the separate `jewel-ui` stage), so the image is never pulled and no layer is created.

File: tools/docker/Dockerfile (hunk 8 above)
Comment (Dostonbek1) [reply tools/docker/Dockerfile:105]: No longer applicable — the `INCLUDE_UI` build arg was removed entirely. The Dockerfile now uses named build targets (`jewel` vs `jewel-ui`), which eliminates the string validation issue.

File: tools/docker/Dockerfile (hunk 9 above)
Comment (Dostonbek1) [reply tools/docker/Dockerfile:106]: No longer applicable — the `cp -r` was removed. The `jewel-ui` stage uses `COPY --from` directly, which copies all files including dotfiles.

**Review by jay-steurer — APPROVED** [review]

File: .github/workflows/build-image.yml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+          persist-credentials: false
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@c94ce9fb468520275223c153574b00df6fe4bcc9 # v3.7.0
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Set up QEMU
```
Comment (tznamena) [inline .github/workflows/build-image.yml:38]: Minor nitpick: it would be nice to have some comments here explaining wh y we need the QEMU and Buildx steps.

**Review by tznamena — APPROVED** [review]
I left a comment recommending to add comments but not those aren't blockers so approving.

File: .github/workflows/build-image.yml (hunk 10 above)
Comment (Dostonbek1) [reply .github/workflows/build-image.yml:38]: Good call — added inline comments explaining that QEMU enables cross-platform emulation (arm64 on amd64 runners) and Buildx is needed for multi-arch builds with direct registry push.

**Comment by Dostonbek1** [comment]
## Test Results

The workflow was tested end-to-end on a fork by temporarily adjusting the repo guard and image name:

- **Workflow run**: ✅ Completed successfully
- **Run URL**: https://github.com/Dostonbek1/jewel/actions/runs/30460913753
- **Image published**: `ghcr.io/dostonbek1/jewel:latest` (+ SHA tags)
- **Multi-arch**: Both `linux/amd64` and `linux/arm64` built and pushed
- **Cleanup**: Fork's `devel` branch restored, test branch deleted

File: tools/docker/Dockerfile [hunk 11]
```diff
@@ -91,11 +91,21 @@ COPY aap_gateway_api /opt/aap_gateway/src/aap_gateway_api/
...
+      '</body></html>' > /opt/aap_gateway/platform_ui/index.html
 
 USER 1000
 EXPOSE 8000
 
 WORKDIR /opt
 CMD ["/usr/bin/launch-gateway"]
+
+FROM jewel AS jewel-ui
+COPY --from=quay.io/ansible/platform-ui:latest /usr/share/nginx/html /opt/aap_gateway/platform_ui
```
[the remark this answers is not in the document]
Comment (Dostonbek1) [reply tools/docker/Dockerfile:111]: Intentionally keeping `:latest` here — the Jinja2 template (`Dockerfile.j2`) used by the local dev environment also references `platform-ui:latest`, and the UI image is updated frequently. Pinning to a digest would add ongoing maintenance burden to manually update it on every UI change. This only affects the optional `jewel-ui` target (not the published GHCR image, which uses the `jewel` target and never pulls this image).
